### PR TITLE
Separated out proposer from attesters

### PR DESCRIPTION
**Pull request by vbuterin**
Also reduces slot duration to 6 seconds and adds a 4-slot delay for attestation inclusion.

### CODE REVIEW

File: specs/core/0_beacon-chain.md [hunk 1]
```diff
@@ -809,12 +838,16 @@ For each one of these attestations:
...
 
-Let `curblock_proposer_index` be the validator index of the `block.slot % len(get_shards_and_committees_for_slot(state, block.slot)[0].committee)`'th attester in `get_shards_and_committees_for_slot(state, block.slot)[0]`, and `parent_proposer_index` be the validator index of the parent block, calculated similarly. Verify that an attestation from the `parent_proposer_index`'th validator is part of the first (ie. item 0 in the array) `AttestationRecord` object; this attester can be considered to be the proposer of the parent block. In general, when a beacon block is produced, it is broadcasted at the network layer along with the attestation from its proposer.
+### Verify proposer signature
 
-Additionally, verify and update the RANDAO reveal. This is done as follows:
+Let `proposal_hash = hash(ProposalSignedData(fork_version, block.slot, 2**64 - 1, block_hash_without_sig))` where `block_hash_without_sig` is the hash of the block except setting `proposer_signature` to `[0, 0]`. 
+
+Verify that `BLSVerify(pubkey=get_proposer(state, block).pubkey, data=proposal_hash, sig=block.proposer_signature)` passes.
```
Comment (djrtwo) [inline specs/core/0_beacon-chain.md]:
```suggestion
Verify that `BLSVerify(pubkey=get_proposer(state, block.slot).pubkey, data=proposal_hash, sig=block.proposer_signature)` passes.
```

* pass in `block.slot` to match the updated `get_beacon_proposer` params

File: specs/core/0_beacon-chain.md [hunk 2]
```diff
@@ -809,12 +838,16 @@ For each one of these attestations:
...
+
+### Verify and process RANDAO reveal
 
 * Let `repeat_hash(x, n) = x if n == 0 else repeat_hash(hash(x), n-1)`.
-* Let `V = state.validators[curblock_proposer_index]`.
+* Let `V = get_proposer(state, block).
```
Comment (djrtwo) [inline specs/core/0_beacon-chain.md]:
```suggestion
* Let `V = get_beacon_proposer(state, block.slot).
```

File: specs/core/0_beacon-chain.md [hunk 3]
```diff
@@ -863,6 +896,19 @@ Perform the following checks:
...
+```python
+{
+    'proposer_index': 'uint24',
+    'proposal1_data': ProposalSignedData,
+    'proposal1_sig': '[uint256]',
+    'proposal2_data': ProposalSignedData,
+    'proposal1_sig': '[uint256]',
+}
+```
+For each `proposal_sig`, verify that `BLSVerify(pubkey=validators[proposer_index].pubkey, msg=proposal_data, sig=proposal_sig)` passes. Verify that `proposal1_data.slot == proposal2_data.slot` but `proposal1 != proposal2`. If `state.validators[proposer_index].status` does not equal `PENALIZED`, then run `exit_validator(proposer_index, state, penalize=True, current_slot=block.slot)`
```
Comment (djrtwo) [inline specs/core/0_beacon-chain.md]:
```suggestion
For each `proposal_sig`, verify that `BLSVerify(pubkey=validators[proposer_index].pubkey, msg=hash(proposal_data), sig=proposal_sig)` passes. Verify that `proposal1_data.slot == proposal2_data.slot` but `proposal1 != proposal2`. If `state.validators[proposer_index].status` does not equal `PENALIZED`, then run `exit_validator(proposer_index, state, penalize=True, current_slot=block.slot)`
```

Need to hash `proposal_data`

File: specs/core/0_beacon-chain.md [hunk 4]
```diff
@@ -38,15 +38,16 @@ The primary source of load on the beacon chain are "attestations". Attestations
 | `GWEI_PER_ETH` | 10**9 | Gwei/ETH |
 | `DEPOSIT_CONTRACT_ADDRESS` | **TBD** | - |
 | `TARGET_COMMITTEE_SIZE` | 2**8 (= 256) | validators |
+| `MIN_ATTESTATION_INCLUSION_DELAY` | 2**2 (= 4) | slots | ~24 seconds |
```
Comment (hwwhww) [inline specs/core/0_beacon-chain.md]: What do you think about moving this after `SLOT_DURATION`?

File: specs/core/0_beacon-chain.md (hunk 3 above)
Comment (hwwhww) [inline specs/core/0_beacon-chain.md]: Naming: unify `proposal_sig` and `proposer_signature`?

File: specs/core/0_beacon-chain.md [hunk 5]
```diff
@@ -486,6 +505,15 @@ def get_block_hash(state: BeaconState,
 
 `get_block_hash(_, _, s)` should always return the block hash in the beacon chain at slot `s`, and `get_shards_and_committees_for_slot(_, s)` should not change unless the validator set changes.
 
+The following is a function that determines the proposer of a block:
+
+```python
+def get_proposer(state, block):
```
[the remark this answers is not in the document]
Comment (vbuterin) [reply specs/core/0_beacon-chain.md]: Fixed.

File: specs/core/0_beacon-chain.md [hunk 6]
```diff
@@ -105,6 +107,8 @@ A `BeaconBlock` has the following fields:
     'slot': 'uint64',
     # Proposer RANDAO reveal
     'randao_reveal': 'hash32',
+    # Proposer signature
+    'proposer_signature': ['uint256'],
```
[the remark this answers is not in the document]
Comment (vbuterin) [reply specs/core/0_beacon-chain.md]: Fixed.

File: specs/core/0_beacon-chain.md (hunk 1 above)
Comment (vbuterin) [reply specs/core/0_beacon-chain.md]: Fixed.

File: specs/core/0_beacon-chain.md (hunk 2 above)
Comment (vbuterin) [reply specs/core/0_beacon-chain.md]: Fixed.

File: specs/core/0_beacon-chain.md (hunk 3 above)
Comment (vbuterin) [reply specs/core/0_beacon-chain.md]: Changed sig -> signature, but I think "proposal" should stay as it's talking about two signatures of two different proposals, but from the same proposer.

File: specs/core/0_beacon-chain.md [hunk 7]
```diff
@@ -147,6 +151,21 @@ An `AttestationRecord` has the following fields:
...
+A `ProposalSignedData` has the following fields:
+
+```python
+{
+    # Fork version
+    'fork_version': 'uint64',
+    # Slot number
+    'slot': 'uint64',
+    # Shard ID (or `2**64 - 1` for beacon chain)
+    'shard_id': 'uint64',
```
Comment (hwwhww) [inline specs/core/0_beacon-chain.md]: Note: the `shard_id`: `uint64` change is in #145.

File: specs/core/0_beacon-chain.md [hunk 8]
```diff
@@ -491,6 +510,15 @@ def get_block_hash(state: BeaconState,
 
 `get_block_hash(_, _, s)` should always return the block hash in the beacon chain at slot `s`, and `get_shards_and_committees_for_slot(_, s)` should not change unless the validator set changes.
 
+The following is a function that determines the proposer of a beacon block:
+
+```python
+def get_beacon_proposer(state, slot):
```
Comment (hwwhww) [inline specs/core/0_beacon-chain.md]:
```suggestion
def get_beacon_proposer(state:BeaconState, slot: int) -> ValidatorRecord:
```

File: specs/core/0_beacon-chain.md [hunk 9]
```diff
@@ -491,6 +510,15 @@ def get_block_hash(state: BeaconState,
...
+    first_committee = get_shards_and_committees_for_slot(state, slot)[0]
```
Comment (hwwhww) [inline specs/core/0_beacon-chain.md]: Note that in Python implement, we'll handle the case when `get_shards_and_committees_for_slot(state, slot)` is an empty list.

Comment (djrtwo) [reply specs/core/0_beacon-chain.md]: can't ever be empty unless the chain stops! but then there will be no proposers :)

**Pull request merged**